% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/writeMeltedMatrixToDisk.R
\name{writeMeltedChromosomalMatrixToDisk}
\alias{writeMeltedChromosomalMatrixToDisk}
\title{Write a matrix, with genes, of a submatrix of a whole genome interaction matrix to disk.}
\usage{
writeMeltedChromosomalMatrixToDisk(whole_genome_matrix, chrom1, chrom2,
  filename, extra_data_matrix = NULL, transpose = F, sequential = T,
  debug = T, multipass = T, desired_range_start = 50,
  desired_range_end = 300)
}
\arguments{
\item{whole_genome_matrix}{A matrix to have edges averaged with genomic coordinates in the form chr1_50_100 set as the column and row names.}

\item{chrom1}{first chromosome of the two which will subset the matrix. (this is done in row-column fasion).}

\item{chrom2}{second chromosome of the two which will subset the matrix. (this is done in row-column fasion).}

\item{extra_data_matrix}{A matrix with additional variables about each point, one position per row with as many variables as remaining columns.}
}
\value{
ggplotmatrix a matrix with values sufficient to create a ggplot2 heatmap with geom_tile() or with ggiraph's geom_tile_interactive()
}
\description{
Writes an RData file with a ggplot2 object within.
}
\keyword{genomic}
\keyword{ggiraph}
\keyword{ggplot2}
\keyword{heatmaply}
\keyword{matrix}
\keyword{plotly}
